nexus/db-queries/src/db/datastore/ip_pool.rs

```diff
@@ -65,44 +63,78 @@ impl DataStore {
         .await
         .map_err(|e| public_error_from_diesel_pool(e, ErrorHandler::Server))
     }
 
-    /// Looks up the default IP pool by name.
-    pub async fn ip_pools_fetch_default_for(
+    /// Look up the default IP pool for the current silo. If there is no default
+    /// at silo scope, fall back to the next level up, namely the fleet default.
+    /// There should always be a default pool at the fleet level, though this
+    /// query can theoretically fail if someone is able to delete that pool or
+    /// make another one the default and delete that.
+    pub async fn ip_pools_fetch_default(
         &self,
         opctx: &OpContext,
-        action: authz::Action,
-    ) -> LookupResult<(authz::IpPool, IpPool)> {
-        self.ip_pools_fetch_for(
-            opctx,
-            action,
-            &Name(ExternalName::from_str("default").unwrap()),
-        )
-        .await
+    ) -> LookupResult<IpPool> {
+        use db::schema::ip_pool::dsl;
+
+        let authz_silo_id = opctx.authn.silo_required()?.id();
+
+        // TODO: Need auth check here. Only fleet viewers can list children on
+        // IP_POOL_LIST, so if we check that, nobody can make instances. This
+        // used to check CreateChild on an individual IP pool, but now we're not
+        // looking up by name so the check is more complicated
+        //
+        // opctx
+        //     .authorize(authz::Action::ListChildren, &authz::IP_POOL_LIST)
+        //     .await?;
+
+        dsl::ip_pool
+            .filter(dsl::silo_id.eq(authz_silo_id).or(dsl::silo_id.is_null()))
+            .filter(dsl::is_default.eq(true))
+            .filter(dsl::time_deleted.is_null())
+            // this will sort by most specific first, i.e.,
+            //
+            //   (silo)
+            //   (null)
+            //
+            // then by only taking the first result, we get the most specific one
+            .order(dsl::silo_id.asc().nulls_last())
+            .select(IpPool::as_select())
+            .first_async::<IpPool>(self.pool_authorized(opctx).await?)
+            .await
+            .map_err(|e| public_error_from_diesel_pool(e, ErrorHandler::Server))
     }
 
-    /// Looks up an IP pool by name.
-    pub(crate) async fn ip_pools_fetch_for(
+    /// Looks up an IP pool by name if it does not conflict with your current scope.
+    pub(crate) async fn ip_pools_fetch(
         &self,
         opctx: &OpContext,
-        action: authz::Action,
         name: &Name,
-    ) -> LookupResult<(authz::IpPool, IpPool)> {
+    ) -> LookupResult<IpPool> {
+        let authz_silo_id = opctx.authn.silo_required()?.id();
+
         let (.., authz_pool, pool) = LookupPath::new(opctx, &self)
             .ip_pool_name(&name)
-            .fetch_for(action)
+            // any authenticated user can CreateChild on an IP pool. this is
+            // meant to represent allocating an IP
+            .fetch_for(authz::Action::CreateChild)
```

Comment:
Why get rid of the `action` arg and hardcode `CreateChild` here? Makes it a little mislead as the method is named `ip_pools_fetch` but as you noted, the `CreateChild` action is for allocating an IP from the pool. So a caller that just wants to retrieve the pool but not allocate from it has a slightly stronger than necessary check.

Reply:
I think the abstraction is just cut the wrong way here. I'm not sure whether I made it worse — it was already weird before. Both this auth check and the check against the silo scope are fundamentally about the use of this function for retrieving an IP pool in order to allocate IPs from it. Despite the name, this isn't a general-purpose function at all. It only has one non-test call site:

https://github.com/oxidecomputer/omicron/blob/01e730acf3efeb6ff2181d558cf373ded383a365/nexus/db-queries/src/db/datastore/external_ip.rs#L47-L68

I'm thinking the thing to do is inline this logic in that call site. `ip_pool_fetch_default` function involves a DB query so it makes sense as a datastore function, plus the `default` imples a "relative to what?" that I think makes the silo ID check make more sense. Plus the query needs to know about the silo ID in order to work.